File: Data Structures & Algorithms/swim-in-rising-water/submission-3.py
from heapq import heappush, heappop

class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        dist = [[float("inf") for _ in range(rows)] for _ in range(cols)]
        dist[0][0] = grid[0][0]

        minHeap = [(grid[0][0], 0, 0)] # (d, r, c)

        while minHeap:
            d, r, c = heappop(minHeap)
            if r == rows - 1 and c == cols - 1:
                return d
            if dist[r][c] < d:
                continue
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols:
                    next_d = max(d, grid[nr][nc])
                    if next_d < dist[nr][nc]:
                        dist[nr][nc] = next_d
                        heappush(minHeap, (next_d, nr, nc))
    

File: Data Structures & Algorithms/max-area-of-island/submission-3.java
class Solution {
    
    private static final int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

    public int maxAreaOfIsland(int[][] grid) {
        int rows = grid.length, cols = grid[0].length;

        int maxArea = 0;
        for(int r = 0; r < rows; r++){
            for(int c = 0; c < cols; c++){
                if(grid[r][c] == 1)
                    maxArea = Math.max(maxArea, dfs(grid, r, c));
            }
        }
        return maxArea;
    }

    public int dfs(int[][] grid, int r, int c){
        int rows = grid.length, cols = grid[0].length;
        grid[r][c] = 0;
        int area = 1;
        for(int[] dir: directions){
            int nr = r + dir[0], nc = c + dir[1];
            if(
                (nr >= 0 && nr < rows)
                && (nc >= 0 && nc < cols)
                && (grid[nr][nc] == 1)
            )
                area += dfs(grid, nr, nc);
        }
        return area;
    }
}
